west: runners: bossac: Honor --erase flag

petejohanson opened the pull request:
Instead of unconditionally erasing the whole target, add support for using the common --erase flag.

This tweak is needed to avoid the bossac runner from erasing the whole chip every time it's run, helping preserve things like settings stored to flash when using them on samd21, for example.

petejohanson committed:
west: runners: bossac: Honor --erase flag

Instead of unconditionally erasing the whole target, add
support for using the common --erase flag.

Signed-off-by: Peter Johanson <peter@peterjohanson.com>

pdgendt reviewed (commented):
I agree with the change, but the default behavior changes now. This should be added to the migration guide.

petejohanson commented:
> I agree with the change, but the default behavior changes now. This should be added to the migration guide.

Will do. Also noticed the test failure, will work on fixing that up.

petejohanson commented:
> I agree with the change, but the default behavior changes now. This should be added to the migration guide.

Added a note to the migration guide, fixed the existing tests, and added a new one for the `--erase` option.

fabiobaltieri commented:
@nandojve should the bossac runner be under SAM in the maintaner file?

nandojve commented:
Hi @petejohanson ,

All the changes on BOSSAC will be carefully checked and this takes times due to complexity around.

@fabiobaltieri ,

> should the bossac runner be under SAM in the maintaners file?

makes sense.

nandojve commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -180,9 +181,12 @@ def magic_delay(self):
     def make_bossac_cmd(self):
         self.ensure_output('bin')
-        cmd_flash = [self.bossac, '-p', self.port, '-R', '-e', '-w', '-v',
+        cmd_flash = [self.bossac, '-p', self.port, '-R', '-w', '-v',
                      '-b', self.cfg.bin_file]
 
+        if self.erase:
+            cmd_flash += ['-e']
```
Would this change making all the boards in tree and out of tree not flash anymore ?

petejohanson replied:
They would still flash the given bin, just not erase the entire flash of the target.

nandojve replied:
Perfect!

Please, make sure this is clear on BOSSA docs.

petejohanson replied:
I've added a note in the host-tools docs about this. Thanks.

nandojve commented on scripts/west_commands/runners/bossac.py:
```diff
@@ -180,9 +181,12 @@ def magic_delay(self):
     def make_bossac_cmd(self):
         self.ensure_output('bin')
-        cmd_flash = [self.bossac, '-p', self.port, '-R', '-e', '-w', '-v',
+        cmd_flash = [self.bossac, '-p', self.port, '-R', '-w', '-v',
```
I don't understand why you are using the flash command if you intention is not to flash. On my perspective there is nothing wrong with the erase here and it should always erase the device before flash.

Would be the case to have a different command on west that fits with your needs ?

petejohanson replied:
The intention is to flash the program, just not to erase the entire flash of the target. Many of the other flashing tools behave this way, so that flash ranges used by the application to preserve settings don't lose their settings, as you are iterating flashing the code.

If you need a full erase, you can pass the  `--erase` flag.

nandojve reviewed (changes requested):
Hi @petejohanson ,

In the meantime, you can check and update documentation here.
https://github.com/zephyrproject-rtos/zephyr/blob/7271000fe56c75536831741d46ef225d08a34f3a/doc/develop/flash_debug/host-tools.rst?plain=1#L17

A few questions about the use of this.
My main question is what you need to do exaclty?

nandojve commented:
Hi @petejohanson ,

Just do mention that the bootloader developed to SAM4L needs a "full erase", which is not an issue. This board will be a reference #83310.
